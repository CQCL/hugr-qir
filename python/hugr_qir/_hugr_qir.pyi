def cli(*args: str) -> None: ...
def opt_level_choices() -> list[str]: ...
def compile_target_choices() -> list[str]: ...
def output_format_choices() -> list[str]: ...
