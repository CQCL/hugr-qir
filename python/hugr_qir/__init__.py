def main_cli():
    from ._hugr_qir import cli
    import sys
    cli(*sys.argv)
