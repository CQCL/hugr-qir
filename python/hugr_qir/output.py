from base64 import b64encode
from enum import Enum
from pathlib import Path

from llvmlite.binding import (  # type: ignore[import-untyped]
    create_context,
    parse_assembly,
)


class OutputFormat(Enum):
    LLVMIR = "llvmir"
    BITCODE = "bitcode"
    BASE64 = "base64"


def get_write_mode(out_format: OutputFormat | None, file_path: Path | None) -> str:
    if out_format and out_format in [OutputFormat.BITCODE, OutputFormat.BASE64]:
        return "wb"
    if file_path and file_path.suffix not in [".ll", ".asm"]:
        return "wb"
    return "w"


def ir_string_to_output_format(qir_ir: str, output_format: OutputFormat) -> str | bytes:
    match output_format:
        case OutputFormat.LLVMIR:
            return qir_ir
        case OutputFormat.BITCODE:
            ctx = create_context()
            module = parse_assembly(qir_ir, context=ctx)
            return module.as_bitcode()
        case OutputFormat.BASE64:
            ctx = create_context()
            module = parse_assembly(qir_ir, context=ctx)
            qir_bitcode = module.as_bitcode()
            return b64encode(qir_bitcode).decode("utf-8")
    errmsg = "Unrecognized output format"
    raise ValueError(errmsg)
