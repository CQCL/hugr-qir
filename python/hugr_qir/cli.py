"""Cli for hugr-qir."""

import tempfile
from importlib.metadata import version
from pathlib import Path

import click
from quantinuum_qircheck import qircheck

from hugr_qir._hugr_qir import cli


@click.command(name="hugr-qir")
@click.argument("hugr_file", type=click.Path(exists=True, path_type=Path))
@click.option(
    "--validate-qir/--no-validate-qir",
    "validate",
    default=True,
    help="Whether to validate output",
)
@click.option(
    "-o",
    "--output",
    "outfile",
    type=click.File("w"),
    default="-",
    help="Name of output file (optional)",
)
@click.version_option(version=version("hugr_qir"))
def hugr_qir(validate: bool, hugr_file: Path, outfile: click.File) -> None:
    """This is the cli for converting hugr to qir."""

    options = ["-q"]
    with tempfile.NamedTemporaryFile(delete=True, suffix=".ll") as temp_file:
        tmp_options = [*options, "-o", temp_file.name]
        cli(str(hugr_file), *tmp_options)
        with Path.open(Path(temp_file.name)) as output:
            qir = output.read()
    if validate:
        qircheck(qir)

    if hasattr(outfile, "write"):
        outfile.write(qir)
    else:
        # this should never happen but the check for write is
        # needed to appease mypy (click doesn't handle typing well)
        err_msg = f"Could not write to object {outfile}"
        raise AttributeError(err_msg)


if __name__ == "__main__":
    hugr_qir()
