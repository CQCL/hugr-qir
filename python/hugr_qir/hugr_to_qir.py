import tempfile
from pathlib import Path

from .cli import hugr_qir_impl


def hugr_to_qir(hugr: bytes, validate_qir: bool = True) -> str:
    """A function for converting hugr to qir."""
    with (
        tempfile.NamedTemporaryFile(delete=True, suffix=".hugr") as temp_infile,
        tempfile.NamedTemporaryFile(delete=True, suffix=".ll") as temp_outfile,
    ):
        with Path.open(Path(temp_infile.name), "wb") as cli_input:
            cli_input.write(hugr)
        with Path.open(Path(temp_outfile.name), "w") as cli_output:
            hugr_qir_impl(validate_qir, Path(cli_input.name), cli_output)
        with Path.open(Path(temp_outfile.name), "r") as cli_output:
            return cli_output.read()
