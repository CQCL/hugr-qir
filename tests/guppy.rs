use std::{env, path::Path, process::{self, Command, Stdio}};

use hugr::{package::Package, std_extensions::STD_REG, Hugr};
use hugr_llvm::inkwell;
use hugr_qir::CompileArgs;
use insta::assert_snapshot;
use itertools::Itertools as _;
use rstest::rstest;

fn capture_guppy(path: impl AsRef<Path>) -> (Hugr, String) {
    let process::Output {
        status,
        stdout,
        stderr,
    } = Command::new(".devenv/state/venv/bin/python")
        .arg(path.as_ref())
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .unwrap();
    eprintln!("{}", std::str::from_utf8(&stdout).unwrap());
    eprintln!("{}", std::str::from_utf8(&stderr).unwrap());
    assert!(status.success());
    let hugr = Package::from_json_reader(stdout.as_slice(), &STD_REG)
        .unwrap()
        .modules
        .into_iter()
        .exactly_one()
        .unwrap();
    let stderr = String::from_utf8(stderr).unwrap();
    (hugr, stderr)
}

fn compile(hugr: &mut Hugr) -> String {
    let args = {
        let mut a = CompileArgs::default();
        a.validate = true;
        a
    };
    let context = inkwell::context::Context::create();
    let module = args.compile(hugr, &context).unwrap();
    module.to_string()
}

// This generates a test case for each file matching 'guppy_examples/**/*.py'
//
// For now we snapshot test the stderr output of executing the python file,
// and the LLVM IR output by hugr-qir.
//
// If we get problems with new test cases not being run in dev environments we might consider
// adding a `build.rs` as in:
// <https://docs.rs/rstest/latest/rstest/attr.rstest.html#files-path-as-input-arguments>
#[rstest]
fn guppy_examples(
    #[base_dir = "guppy_examples"]
    #[files("**/*.py")]
    file: impl AsRef<Path>,
) {
    let file2 = file.as_ref();
    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_suffix(file2.file_stem().unwrap().to_str().unwrap());
    settings.bind(|| {
        let (mut hugr, stderr) = capture_guppy(file);
        assert_snapshot!("llvmir", compile(&mut hugr));
        assert_snapshot!("stderr", stderr);
    });
}
