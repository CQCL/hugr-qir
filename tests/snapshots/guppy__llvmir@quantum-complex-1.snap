---
source: tests/guppy.rs
expression: compile(&mut hugr)
---
; ModuleID = 'hugr-qir'
source_filename = "hugr-qir"

%QUBIT = type opaque
%RESULT = type opaque

@0 = private unnamed_addr constant [47 x i8] c"tket2.rotation.from_halfturns_unchecked failed\00", align 1
@prelude.panic_template = private unnamed_addr constant [34 x i8] c"Program panicked (signal %i): %s\0A\00", align 1
@1 = private unnamed_addr constant [47 x i8] c"tket2.rotation.from_halfturns_unchecked failed\00", align 1
@prelude.panic_template.1 = private unnamed_addr constant [34 x i8] c"Program panicked (signal %i): %s\0A\00", align 1
@2 = private unnamed_addr constant [47 x i8] c"tket2.rotation.from_halfturns_unchecked failed\00", align 1
@prelude.panic_template.2 = private unnamed_addr constant [34 x i8] c"Program panicked (signal %i): %s\0A\00", align 1
@3 = private unnamed_addr constant [37 x i8] c"Expected variant 1 but got variant 0\00", align 1
@prelude.panic_template.3 = private unnamed_addr constant [34 x i8] c"Program panicked (signal %i): %s\0A\00", align 1
@4 = private unnamed_addr constant [37 x i8] c"Expected variant 1 but got variant 0\00", align 1
@prelude.panic_template.4 = private unnamed_addr constant [34 x i8] c"Program panicked (signal %i): %s\0A\00", align 1
@5 = private unnamed_addr constant [2 x i8] c"1\00", align 1
@6 = private unnamed_addr constant [2 x i8] c"0\00", align 1

define void @__hugr__.main.1() #0 {
alloca_block:
  br i1 true, label %2, label %0

0:                                                ; preds = %alloca_block
  %1 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([34 x i8], [34 x i8]* @prelude.panic_template, i32 0, i32 0), i32 1, i8* getelementptr inbounds ([47 x i8], [47 x i8]* @0, i32 0, i32 0))
  call void @abort()
  unreachable

2:                                                ; preds = %alloca_block
  br i1 true, label %5, label %3

3:                                                ; preds = %2
  %4 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([34 x i8], [34 x i8]* @prelude.panic_template.1, i32 0, i32 0), i32 1, i8* getelementptr inbounds ([47 x i8], [47 x i8]* @1, i32 0, i32 0))
  call void @abort()
  unreachable

5:                                                ; preds = %2
  br i1 true, label %8, label %6

6:                                                ; preds = %5
  %7 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([34 x i8], [34 x i8]* @prelude.panic_template.2, i32 0, i32 0), i32 1, i8* getelementptr inbounds ([47 x i8], [47 x i8]* @2, i32 0, i32 0))
  call void @abort()
  unreachable

8:                                                ; preds = %5
  br i1 true, label %cond_278_case_1, label %cond_278_case_0

cond_278_case_1:                                  ; preds = %8
  br label %cond_exit_278

cond_278_case_0:                                  ; preds = %8
  %9 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([34 x i8], [34 x i8]* @prelude.panic_template.3, i32 0, i32 0), i32 1, i8* getelementptr inbounds ([37 x i8], [37 x i8]* @3, i32 0, i32 0))
  call void @abort()
  br label %cond_exit_278

cond_exit_278:                                    ; preds = %cond_278_case_1, %cond_278_case_0
  br i1 true, label %cond_291_case_1, label %cond_291_case_0

cond_291_case_1:                                  ; preds = %cond_exit_278
  br label %cond_exit_291

cond_291_case_0:                                  ; preds = %cond_exit_278
  %10 = call i32 (i8*, ...) @printf(i8* getelementptr inbounds ([34 x i8], [34 x i8]* @prelude.panic_template.4, i32 0, i32 0), i32 1, i8* getelementptr inbounds ([37 x i8], [37 x i8]* @4, i32 0, i32 0))
  call void @abort()
  br label %cond_exit_291

cond_exit_291:                                    ; preds = %cond_291_case_1, %cond_291_case_0
  %"019.0.reg2mem.0" = phi %QUBIT* [ inttoptr (i64 1 to %QUBIT*), %cond_291_case_1 ], [ null, %cond_291_case_0 ]
  call void @__quantum__qis__phasedx__body(double 0x3FF921FB54442D18, double 0xBFF921FB54442D18, %QUBIT* %"019.0.reg2mem.0")
  call void @__quantum__qis__rz__body(double 0x400921FB54442D18, %QUBIT* %"019.0.reg2mem.0")
  call void @__quantum__qis__rz__body(double 0x3FF921FB54442D18, %QUBIT* %"019.0.reg2mem.0")
  call void @__quantum__qis__rz__body(double 0x3FE921FB54442D18, %QUBIT* %"019.0.reg2mem.0")
  call void @__quantum__qis__phasedx__body(double 0x400921FB54442D18, double 0.000000e+00, %QUBIT* %"019.0.reg2mem.0")
  call void @__quantum__qis__phasedx__body(double 0x400921FB54442D18, double 0x3FF921FB54442D18, %QUBIT* %"019.0.reg2mem.0")
  call void @__quantum__qis__rz__body(double 0x400921FB54442D18, %QUBIT* %"019.0.reg2mem.0")
  call void @__quantum__qis__phasedx__body(double 0xBFF921FB54442D18, double 0x3FF921FB54442D18, %QUBIT* null)
  call void @__quantum__qis__zzmax__body(%QUBIT* %"019.0.reg2mem.0", %QUBIT* null)
  call void @__quantum__qis__rz__body(double 0xBFF921FB54442D18, %QUBIT* %"019.0.reg2mem.0")
  call void @__quantum__qis__phasedx__body(double 0x3FF921FB54442D18, double 0x400921FB54442D18, %QUBIT* null)
  call void @__quantum__qis__rz__body(double 0xBFF921FB54442D18, %QUBIT* null)
  call void @__quantum__qis__phasedx__body(double 0x400921FB54442D18, double 0x400921FB54442D18, %QUBIT* %"019.0.reg2mem.0")
  call void @__quantum__qis__zzmax__body(%QUBIT* %"019.0.reg2mem.0", %QUBIT* null)
  call void @__quantum__qis__rz__body(double 0x3FF921FB54442D18, %QUBIT* null)
  call void @__quantum__qis__phasedx__body(double 0x400921FB54442D18, double 0x3FF921FB54442D18, %QUBIT* %"019.0.reg2mem.0")
  call void @__quantum__qis__rz__body(double 0xBFF921FB54442D18, %QUBIT* %"019.0.reg2mem.0")
  %11 = call %RESULT* @__quantum__qis__m__body(%QUBIT* null)
  %12 = call i1 @__quantum__qis__read_result__body(%RESULT* %11)
  call void @__quantum__qis__phasedx__body(double 0x4012D97C7F3321D2, double 0.000000e+00, %QUBIT* %"019.0.reg2mem.0")
  call void @__quantum__qis__phasedx__body(double 0x4012D97C7F3321D2, double 0x3FF921FB54442D18, %QUBIT* %"019.0.reg2mem.0")
  call void @__quantum__qis__rz__body(double 0x4012D97C7F3321D2, %QUBIT* %"019.0.reg2mem.0")
  %13 = call %RESULT* @__quantum__qis__m__body(%QUBIT* %"019.0.reg2mem.0")
  %14 = call i1 @__quantum__qis__read_result__body(%RESULT* %13)
  call void @__quantum__rt__bool_record_output(i1 %12, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @5, i32 0, i32 0))
  call void @__quantum__rt__bool_record_output(i1 %14, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @6, i32 0, i32 0))
  ret void
}

declare i32 @printf(i8*, ...)

declare void @abort()

declare %QUBIT* @__quantum__rt__qubit_allocate()

declare void @__quantum__qis__phasedx__body(double, double, %QUBIT*)

declare void @__quantum__qis__rz__body(double, %QUBIT*)

declare void @__quantum__qis__zzmax__body(%QUBIT*, %QUBIT*)

declare %RESULT* @__quantum__qis__m__body(%QUBIT*)

declare i1 @__quantum__qis__read_result__body(%RESULT*)

declare void @__quantum__rt__bool_record_output(i1, i8*)

attributes #0 = { "entry_point" "output_labeling_schema" "qir_profiles"="custom" "required_num_qubits"="20" "required_num_results"="20" }

!llvm.module.flags = !{!0, !1, !2, !3}

!0 = !{i32 1, !"qir_major_version", i32 1}
!1 = !{i32 7, !"qir_minor_version", i32 0}
!2 = !{i32 1, !"dynamic_qubit_management", i1 false}
!3 = !{i32 1, !"dynamic_result_management", i1 false}
